pkg/cmd/cli/cmd/loginoptions.go

```diff
@@ -71,122 +36,178 @@ func (o *LoginOptions) GatherInfo() error {
-		o.Username = me.Name
-		fmt.Printf("Already logged into %q as %q.\n", o.Config.Host, o.Username)
-
-	} else {
-		// if not, we need to log in again
-
-		o.Config.BearerToken = ""
-		o.Config.CertFile = o.CertFile
-		o.Config.CertData = o.CertData
-		o.Config.KeyFile = o.KeyFile
-		o.Config.KeyData = o.KeyData
-		token, err := tokencmd.RequestToken(o.Config, o.Reader, o.Username, o.Password)
-		if err != nil {
-			return err
+	// make a copy and use it to avoid mutating the original
+	t := *directClientConfig
+	clientConfig := &t
+
+	// make sure we have a username before continuing
+	if !o.usernameProvided() {
+		if cmdutil.IsTerminal(o.Reader) {
+			for !o.usernameProvided() {
+				o.Username = cmdutil.PromptForString(o.Reader, "Username: ")
+			}
 		}
-		o.Config.BearerToken = token
+	}
 
-		me, err := o.whoAmI()
-		if err != nil {
-			return err
+	// search all valid contexts with matching server stanzas to see if we have a matching user stanza
+	kubeconfig := *o.StartingKubeConfig
+	matchingClusters := getMatchingClusters(*clientConfig, kubeconfig)
+
+	for key, context := range o.StartingKubeConfig.Contexts {
+		if matchingClusters.Has(context.Cluster) {
+			clientcmdConfig := kclientcmd.NewDefaultClientConfig(kubeconfig, &kclientcmd.ConfigOverrides{CurrentContext: key})
+			if kubeconfigClientConfig, err := clientcmdConfig.ClientConfig(); err == nil {
+				if osClient, err := client.New(kubeconfigClientConfig); err == nil {
+					if me, err := whoAmI(osClient); err == nil && (o.Username == me.Name) {
+						o.Config.BearerToken = kubeconfigClientConfig.BearerToken
+						o.Config.CertFile = kubeconfigClientConfig.CertFile
+						o.Config.CertData = kubeconfigClientConfig.CertData
+						o.Config.KeyFile = kubeconfigClientConfig.KeyFile
+						o.Config.KeyData = kubeconfigClientConfig.KeyData
+
+						if key == o.StartingKubeConfig.CurrentContext {
+							fmt.Printf("Already logged into %q as %q.\n", o.Config.Host, o.Username)
+						}
+						return nil
+					}
+				}
+			}
 		}
-		o.Username = me.Name
-		fmt.Println("Login successful.\n")
 	}
 
-	// TODO investigate about the safety and intent of the proposal below
-	// if trying to log in an user that's not the currently logged in, try to reuse an existing token
-
-	// if o.usernameProvided() {
-	// 	glog.V(5).Infof("Checking existing authentication info for '%v'...\n", o.Username)
-
-	// 	for _, ctx := range rawCfg.Contexts {
-	// 		authInfo := rawCfg.AuthInfos[ctx.AuthInfo]
-	// 		clusterInfo := rawCfg.Clusters[ctx.Cluster]
-
-	// 		if ctx.AuthInfo == o.Username && clusterInfo.Server == o.Server && len(authInfo.Token) > 0 { // only token for now
-	// 			glog.V(5).Infof("Authentication exists for '%v' on '%v', trying to use it...\n", o.Server, o.Username)
-
-	// 			o.Config.BearerToken = authInfo.Token
+	// if kubeconfig doesn't already have a matching user stanza...
+	clientConfig.BearerToken = ""
+	clientConfig.CertFile = o.CertFile
```

Comment:
Make sure there's no scenario where client config have both ` CertFile` *and* `CertData` or `KeyFile` *and* `KeyData`. That produces an error. You may want to clean `*Data` just in case.